File: python_oop_scraping_bigram/markov.py
import random

class MarkovChain:

    def __init__(self):
        self.chain = {"_start": [], "_end": []}

    def add_phrase(self, phrase):
        words = phrase.lower().strip().split(" ")

        # Loop through each word, check if it's already added
		# if its added, then get the suffix vector and add the word
		# if it hasn't been added then add the word to the list
		# if its the first or last word then select the _start / _end key
        i = 0
        while i < len(words) - 1:
            bigram = f"{words[i]} {words[i + 1]}"
            # Handle case of first or last bigram in phrase
            if i == 0:
                start_words = self.chain.get("_start")
                start_words.append(bigram)
			# Add word_after (or _end) to chain for this bigram
            word_after = "_end"
            if i < len(words) - 2:
                word_after = words[i + 2]
            bigram = bigram
            if bigram not in self.chain:
                self.chain[bigram] = []
            self.chain[bigram].append(word_after)
            i += 1

    def generate_phrase(self):
        """Generate a Markov phrase"""
        new_phrase = []

		# Select the first bigram
        start_bigram = random.choice(self.chain.get("_start"))
        new_phrase.append(start_bigram)

	    # Keep looping through the words until we
        # can't come up with another word or each an "end" marker
        prev_bigram = start_bigram
        prev_word = start_bigram.split(" ")[1]
        new_word = ""
        num_forks = 0
        while True:
            prev_bigram = prev_bigram
            new_words = self.chain.get(prev_bigram)
            if new_words is None:
                break
            if len(new_words) > 1:
                num_forks += 1
            new_word = random.choice(new_words)
            if new_word == "_end":
                break
            new_phrase.append(new_word)
            prev_bigram = prev_word + " " + new_word
            prev_word = new_word

        return f'{" ".join(new_phrase)} [forks: {num_forks}]'

File: java_bigram/Markov.java
import java.util.Hashtable;
import java.util.Random;
import java.util.Vector;

public class Markov {

	// Hashmap
	static Hashtable<String, Vector<String>> chain;
	static Random rnd;

    public Markov() {
        chain = new Hashtable<String, Vector<String>>();
        // Create the first two entries (k:_start, k:_end)
            chain.put("_start", new Vector<String>());
            chain.put("_end", new Vector<String>());
        rnd = new Random();
    }

	/*
	 * Add words
	 */
	public static void addPhrase(String phrase) {
		// put each word into an array
		String[] words = phrase.split(" ");

		// Loop through each word, check if it's already added
		// if its added, then get the suffix vector and add the word
		// if it hasn't been added then add the word to the list
		// if its the first or last word then select the _start / _end key

		for (int i = 0; i < words.length; i++) {

      // Handle case of first or last word in phrase
			if (i == 0) {
				Vector<String> startWords = chain.get("_start");
				startWords.add(words[i]);
			} else if (i == words.length-1) {
				Vector<String> endWords = chain.get("_end");
				endWords.add(words[i]);
			}

      // Handle any word that has a word after it
      if (i < words.length - 1) {
				Vector<String> suffix = chain.get(words[i]);
        if (suffix == null) {
          // If the word isn't in the chain yet,
          // add key=word,vector=<word after>
					suffix = new Vector<String>();
					suffix.add(words[i+1]);
					chain.put(words[i], suffix);
				} else {
          // Otherwise add <word after> to vector
					suffix.add(words[i+1]);
					chain.put(words[i], suffix);
				}
			}
		}
	}


	/*
	 * Generate a markov phrase
	 */
	public static String generateSentence() {

		// Vector to hold the phrase
		Vector<String> newPhrase = new Vector<String>();

		// String for the next word
		String nextWord = "";

		// Select the first word
		Vector<String> startWords = chain.get("_start");
		int startWordsLen = startWords.size();
		nextWord = startWords.get(rnd.nextInt(startWordsLen));
		newPhrase.add(nextWord);

		// Keep looping through the words until we've reached the end
		while (nextWord.charAt(nextWord.length()-1) != '.') {
			Vector<String> wordSelection = chain.get(nextWord);
      if (wordSelection == null) {
        break;
      }
      int wordSelectionLen = wordSelection.size();
			nextWord = wordSelection.get(rnd.nextInt(wordSelectionLen));
			newPhrase.add(nextWord);
		}

		return String.join(" ", newPhrase);
	}
}